contracts/specs/dai.spec

```diff
@@ -0,0 +1,364 @@
+    init_state axiom balanceSum() == 0;
+}
+
+hook Sstore balanceOf[KEY address a] uint256 balance (uint256 old_balance) STORAGE {
+    havoc balanceSum assuming balanceSum@new() == balanceSum@old() + (balance - old_balance);
+}
+
+function strengthenFor2Addresses(address a1, address a2) {
+    require(balanceSum() >= balanceOf(a1) + balanceOf(a2));
+}
+
+// invariants also check the desired property on the constructor
+invariant balanceSum_equals_totalSupply() balanceSum() == totalSupply() {
+    preserved transfer(address to, uint _) with (env e) {
+        strengthenFor2Addresses(e.msg.sender, to);
+    }
+
+    preserved transferFrom(address from, address to, uint _) with (env e) {
+        strengthenFor2Addresses(from, to);
+    }
+
+    preserved mint(address to, uint _) with (env e) {
+        require(balanceSum() >= balanceOf(to));
+    }
+
+    preserved burn(address from, uint _) with (env e) {
+        require(balanceSum() >= balanceOf(from));
+    }
+}
+
+// Verify that wards behaves correctly on rely
+rule rely(address usr) {
+    env e;
+
+    rely(e, usr);
+
+    assert(wards(usr) == 1, "Rely did not set the wards as expected");
+}
+
+// Verify revert rules on rely
+rule rely_revert(address usr) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    rely@withrevert(e, usr);
+
+    bool revert1 = ward != 1;
+    bool revert2 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Lack of auth did not revert");
+    assert(revert2 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that wards behaves correctly on deny
+rule deny(address usr) {
+    env e;
+
+    deny(e, usr);
+
+    assert(wards(usr) == 0, "Deny did not set the wards as expected");
+}
+
+// Verify revert rules on deny
+rule deny_revert(address usr) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    deny@withrevert(e, usr);
+
+    bool revert1 = ward != 1;
+    bool revert2 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Lack of auth did not revert");
+    assert(revert2 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that balance behaves correctly on transfer
+rule transfer(address to, uint256 value) {
+    env e;
+
+    uint256 senderBalance = balanceOf(e.msg.sender);
+    uint256 toBalance = balanceOf(to);
+    bool senderSameAsTo = e.msg.sender == to;
+
+    require(toBalance + value <= max_uint); // Avoid evaluating the overflow case
+
+    transfer(e, to, value);
+
+    assert(!senderSameAsTo =>
+            balanceOf(e.msg.sender) == senderBalance - value &&
+            balanceOf(to) == toBalance + value,
+            "Transfer did not change balances as expected"
+    );
+
+    assert(senderSameAsTo =>
+            balanceOf(e.msg.sender) == senderBalance &&
+            senderBalance == toBalance,
```

Comment:
I don't think this condition adds any assurance and can therefore be removed.

Reply:
you mean just the part `&& senderBalance == toBalance`? In that case yeah, I think you are right.